design/src/rfcs/rfc0022_error_context_and_compatibility.md

```diff
@@ -0,0 +1,370 @@
+- The source error can be downcasted, which could be a trap for backwards compatibility.
+
+### Hyper 1.0
+
+Hyper is has outlined [some problems they want to address with errors](https://github.com/hyperium/hyper/blob/bd7928f3dd6a8461f0f0fdf7ee0fd95c2f156f88/docs/ROADMAP.md#errors)
+for the coming 1.0 release. To summarize:
+
+- It's difficult to match on specific errors (Hyper 0.x's `Error` relies
+  on `is_x` methods for error matching rather than enum matching).
+- Error reporters duplicate information since the hyper 0.x errors include the display of their error sources
+- `Error::source()` can leak internal dependencies
+
+Opaque Error Sources
+--------------------
+
+There is [discussion in the errors working group](https://github.com/rust-lang/project-error-handling/issues/53)
+about how to avoid leaking internal dependency error types through error source downcasting. One option is to
+create an opaque error wrapping new-type that removes the ability to downcast to the other library's error.
+This, however, can be circumvented via unsafe code, and also breaks the ability for error reporters to
+properly display the error (for example, if the error has backtrace information, that would be
+inaccessible to the reporter).
+
+This situation might improve if the nightly `request_value`/`request_ref`/`provide` functions on
+`std::error::Error` are stabilized, since then contextual information needed for including things
+such as a backtrace could still be retrieved through the opaque error new-type.
+
+This RFC proposes that `Box<dyn Error + Send + Sync + 'static>` is used to refer to errors from libraries.
+An allocation to store this can be avoided if the error source is private, in which case, a reference to it
+can be returned from the `Error::source` implementation. Otherwise, if it's directly exposed as a public enum
+field, then it must be boxed.
+
+Ideally, errors should contain enough useful information on them that downcasting the underlying cause
+is not necessary. Customers needing to downcast the error source should be a last resort, and with the
+understanding that the type could change at a later date with no compile-time guarantees.
+
+Error Proposal
+--------------
+
+Taking a customer's perspective, there are two broad categories of errors:
+
+1. **Actionable:** Errors that can/should influence program flow
+2. **Informative:** Errors that inform that something went wrong, but where
+   it's not useful to match on the error to change program flow
+
+This RFC proposes that a consistent pattern be introduced to cover these two use cases for
+all errors in the public API for the Rust runtime crates and generated client crates.
+
+### Actionable error pattern
+
+Actionable errors must be represented as enums. If an error variant has an error source or additional
+contextual information, it must use a struct that is either embedded inside the enum, or separate where it makes
+sense to have private fields. For example:
```

Comment:
I tried to make this more concise but I may have gone too far

```suggestion
Actionable errors are represented as enums. Struct-like enum variants may contain an error source or additional
context. When private fields are required, a tuple-like enum variant is used to wrap a single struct. For example:
```

Reply:
btw, the "struct-like" and "tuple-like" language is lifted directly from https://doc.rust-lang.org/book/ch18-03-pattern-syntax.html?highlight=struct%20variant#destructuring-enums